src/derivers/cip3Icarus.ts

```diff
@@ -0,0 +1,348 @@
+ * @param bytes - Little-Endian big number in bytes.
+ * @returns PadEnd32Bytes(mod(bytes, 2^256))).
+ */
+const mod2Pow256 = (bytes: Uint8Array): Uint8Array => {
+  return padEnd32Bytes(
+    bigIntToBytes(mod(bytesToBigInt(bytes), BigInt(2) ** BigInt(256))),
+  );
+};
+
+/**
+ * Adds the left to the right.
+ *
+ * @param left - Left hand side Little-Endian big number.
+ * @param right - Right hand side Little-Endian big number.
+ * @returns PadEnd32Bytes(left + right).
+ */
+const add = (left: Uint8Array, right: Uint8Array): Uint8Array => {
+  const added = bytesToBigInt(left) + bytesToBigInt(right);
+  return padEnd32Bytes(bigIntToBytes(added)).slice(0, 32);
+};
+
+/**
+ * Concat tag, key and childIndex.
+ *
+ * @param tag - Key specific tag.
+ * @param key - Key.
+ * @param childIndex - Child index.
+ * @returns PadEnd32Bytes(left + right).
+ */
+const getKeyExtension = (tag: number, key: Uint8Array, childIndex: number) => {
+  return concatBytes([
+    new Uint8Array([tag]),
+    key,
+    numberToUint32(childIndex, true),
+  ]);
+};
+
+type DeriveKeyBaseArgs = { childIndex: number };
+
+type DeriveWithPrivateArgs = DeriveKeyBaseArgs & {
+  parentNode: {
+    privateKeyBytes: Uint8Array;
+    chainCodeBytes: Uint8Array;
+    publicKeyBytes: Uint8Array;
+  };
+  isHardened: boolean;
+};
+
+const Z_TAGS = {
+  normal: 2,
+  hardened: 0,
+};
+
+/**
+ * Derives a private child key.
+ *
+ * Following "Section V. BIP32-ED25519: SPECIFICATION, C.1,2" in https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf.
+ *
+ * @param param1 - The parameters for deriving a child key.
+ * @param param1.parentNode - The parent node containing private key, chain code, and public key.
+ * @param param1.childIndex - The index of the child key.
+ * @param param1.isHardened - Indicates if the child key is hardened.
+ * @returns The derived child key.
+ */
+export const derivePrivateKey = async ({
+  parentNode,
+  childIndex,
+  isHardened,
+}: DeriveWithPrivateArgs) => {
+  // extension = i >= 2^31 ? (0x00||kp||i) : (0x02||Ap||i)
+  const extension = isHardened
+    ? getKeyExtension(
+        Z_TAGS.hardened,
+        parentNode.privateKeyBytes,
+        childIndex + BIP_32_HARDENED_OFFSET,
+      )
+    : getKeyExtension(Z_TAGS.normal, parentNode.publicKeyBytes, childIndex);
+
+  // entropy = Fcp(extension)
+  const entropy = generateEntropy({
+    chainCode: parentNode.chainCodeBytes,
+    extension,
+  });
+
+  const zl = entropy.subarray(0, 32);
+  const zr = entropy.subarray(32);
+
+  const parentKl = parentNode.privateKeyBytes.subarray(0, 32);
+  const parentKr = parentNode.privateKeyBytes.subarray(32);
+
+  // 8[ZL] + kPL
+  const childKl = add(trunc28Mul8(zl), parentKl);
+  // ZR + kPR
+  const childKr = add(zr, mod2Pow256(parentKr));
+  return concatBytes([childKl, childKr]);
+};
+
+type DeriveWithoutPrivateArgs = DeriveKeyBaseArgs & {
+  parentNode: {
+    chainCodeBytes: Uint8Array;
+    publicKeyBytes: Uint8Array;
+  };
+  isHardened: false;
```

Comment:
can't this property be inferred from the child index? The convention (AFAIK) is that child indices below `2^31` are considered "unhardened" and the ones above this threshold are "hardened"

https://github.com/bitcoin/bips/blob/e918b50731397872ad2922a1b08a5a4cd1d6d546/bip-0088.mediawiki#matching

Reply:
I see you are passing to this function the child index that has this threshold already removed. I'd consider passing the unaltered child index to remove the need to pass this "isHardened" flag separately which seems to deviate from other pre-existing key derivation functions' interfaces I checked that seem to take the unaltered child index

Reply:
ah sorry, I checked wrong, indeed the `isHardened` flag seems to be passed also in the pre-existing interfaces: https://github.com/nufi-official/metamask-key-tree/blob/9d52f190e0866a765db1c438754f6843cda9823e/src/derivers/shared.ts#L97 all good then 👍